File: src/front-end/fun-chess-app/src/core/chess/Piece.ts
import Move from "./structs/Move";
import Board from "./Board";
import Cell from "./Cell";
import SpecialMove from "./enums/SpecialMove";
import Pawn from "./pieces/Pawn";
import Bishop from "./pieces/Bishop";
import Rook from "./pieces/Rook";

abstract class Piece {
    private static readonly ChessPieces = new Map<number, Piece>();

    private _value: number;

    public constructor(value: number) {
        Piece.ChessPieces.set(value, this);
        this._value = value;
    }

    public move(board: Board, move: Move): [boolean, SpecialMove] {
        const tuple = this.moveIsValid(board, move)
        if (tuple[0]) {
            board.internalBoard[move.next.index] = board.internalBoard[move.previous.index];
            board.internalBoard[move.previous.index] = Cell.Empty;
        }
        return tuple;
    }

    public abstract moveIsValid(board: Board, move: Move): [boolean, SpecialMove]

    public static toByte(piece: Piece | null): number {
        return piece === null ? 0 : piece._value;
    }

    public static parse(value: number): Piece {
        for (let i = 4; i <= 128; i <<= 1) {
            if ((i & value) === 0) continue;
            return Piece.ChessPieces.get(i)!;
        }
        throw new Error("Byte must contain a piece value.");
    }
}

export default Piece;

File: src/back-end/FunChess/src/FunChess.Core.Chess/Piece.cs
using FunChess.Core.Chess.Enums;
using FunChess.Core.Chess.Pieces;
using FunChess.Core.Chess.Structs;

namespace FunChess.Core.Chess;

public abstract class Piece : IEquatable<Piece>
{
    private static readonly Dictionary<byte, Piece> ChessPieces = new(6);

    public static readonly Piece King = new King();
    public static readonly Piece Queen = new Queen();
    public static readonly Piece Rook = new Rook();
    public static readonly Piece Knight = new Knight();
    public static readonly Piece Bishop = new Bishop();
    public static readonly Piece Pawn = new Pawn();

    protected Piece()
    {
        byte key = (byte)(1 << (2 + ChessPieces.Count));

        Value = key;
        ChessPieces.Add(key, this);
    }
    
    private byte Value { get; }

    public abstract bool Move(Board board, Move move);

    public abstract bool MoveIsValid(Board board, Move move, out SpecialMove specialMove);

    public bool Equals(Piece? other)
    {
        if (other is null)
        {
            return false;
        }
        return Value == other.Value;
    }

    public static explicit operator Piece(byte value)
    {
        for (byte i = 4; i <= 128; i <<= 1)
        {
            if ((i & value) == 0) continue;
            return ChessPieces[i];
        }
        throw new ArgumentException("Byte must contain a piece value.");
    }

    public static explicit operator byte(Piece? value)
    {
        return value == null ? (byte)0 : value.Value;
    }

    public static bool operator ==(Piece? left, Piece? right)
    {
        if (left is null)
        {
            return right is null;
        }
        return left.Equals(right);
    }

    public static bool operator !=(Piece? left, Piece? right)
    {
        if (left is null)
        {
            return right is not null;
        }
        return !left.Equals(right);
    }

    public override bool Equals(object? obj)
    {
        if (obj is not Piece piece) return false;
        return piece.Value == Value;
    }

    public override int GetHashCode() => throw new NotImplementedException();
}